iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java

```diff
@@ -0,0 +1,209 @@
+        binaryConverter = new PrimitiveObjectInspectorConverter.BinaryConverter(argumentOI,
+            PrimitiveObjectInspectorFactory.writableBinaryObjectInspector);
+        break;
+
+      case INT:
+        icebergType = Types.IntegerType.get();
+        intConverter = new PrimitiveObjectInspectorConverter.IntConverter(argumentOI,
+            PrimitiveObjectInspectorFactory.writableIntObjectInspector);
+        break;
+
+      case LONG:
+        icebergType = Types.LongType.get();
+        longConverter = new PrimitiveObjectInspectorConverter.LongConverter(argumentOI,
+            PrimitiveObjectInspectorFactory.writableLongObjectInspector);
+        break;
+
+      case DECIMAL:
+        DecimalTypeInfo decimalTypeInfo = (DecimalTypeInfo) TypeInfoUtils.getTypeInfoFromObjectInspector(argumentOI);
+        icebergType = Types.DecimalType.of(decimalTypeInfo.getPrecision(), decimalTypeInfo.getScale());
+
+        decimalConverter = new PrimitiveObjectInspectorConverter.HiveDecimalConverter(argumentOI,
+            PrimitiveObjectInspectorFactory.writableHiveDecimalObjectInspector);
+        break;
+
+      case FLOAT:
+        icebergType = Types.FloatType.get();
+        floatConverter = new PrimitiveObjectInspectorConverter.FloatConverter(argumentOI,
+            PrimitiveObjectInspectorFactory.writableFloatObjectInspector);
+        break;
+
+      case DOUBLE:
+        icebergType = Types.DoubleType.get();
+        doubleConverter = new PrimitiveObjectInspectorConverter.DoubleConverter(argumentOI,
+            PrimitiveObjectInspectorFactory.writableDoubleObjectInspector);
+        break;
+
+
+      default:
+        throw new UDFArgumentException(
+            " ICEBERG_BUCKET() only takes STRING/CHAR/VARCHAR/BINARY/INT/LONG/DECIMAL/FLOAT/DOUBLE" +
+                " types as first argument, got " + inputType);
+    }
+
+    argumentOI = (PrimitiveObjectInspector) arguments[1];
+    if (!isValid2ndArgumentType(arguments[1])) {
+      throw new UDFArgumentTypeException(1,
+          "ICEBERG_BUCKET() second argument only takes an int types, got " + arguments[1].getTypeName());
+    }
+    numBuckets = ((WritableConstantIntObjectInspector) argumentOI).getWritableConstantValue().get();
+
+    outputOI = PrimitiveObjectInspectorFactory.writableIntObjectInspector;
+    return outputOI;
+  }
+
+  private boolean isValid2ndArgumentType(ObjectInspector arg) {
+    if (arg.getCategory() != ObjectInspector.Category.PRIMITIVE) {
+      return false;
+    } else {
+      PrimitiveObjectInspector.PrimitiveCategory inputType = argumentOI.getPrimitiveCategory();
+      if (inputType != PrimitiveObjectInspector.PrimitiveCategory.INT) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  @Override
+  public Object evaluate(DeferredObject[] arguments) throws HiveException {
+
+    DeferredObject argument = arguments[0];
+    if (argument == null) {
+      return null;
+    }
+
+    if (Types.StringType.get().equals(icebergType)) {
```

Comment:
Can we do this with a method, so we do not have to run the whole if/else misey?

Reply:
Not sure how you mean - originally I wanted a switch, but the Iceberg types are unavailable as enums. But then again for some types the logic is different, like bytes, decimals.

Reply:
Discussed offline - will be refactoring